### PR TITLE
DIG-2267 add samples

### PR DESCRIPTION
**Pull request by SonQBChau**
## What's new
- add upload samples endpoint
- add get samples by dataset endpoint

### How to test
- No unit tests yet, test with candigv3 stack
- Rebuild the stack with this branch
- Ingest [input_mohccn_data.json](https://github.com/user-attachments/files/25755279/input_mohccn_data.json) with prefix SITE_PM2C
```bash
curl --request POST \
  --url 'http://candig.docker.internal:8008/v1/sample/upload?prefix=SITE_PM2C' \
  --header 'accept: application/json' \
  --header 'authorization: Bearer token' \
  --header 'content-type: multipart/form-data' \
  --form file=input_mohccn_data.json
```
- Query /datasets/{dataset_id}/samples
```bash
curl --request GET \
  --url 'http://candig.docker.internal:5080/candig-api/v1/datasets/SITE_PM2C~SYNTH_01/samples' \
  --header 'authorization: Bearer token'
 ```

### CODE REVIEW

File: schema.yml [hunk 1]
```diff
@@ -318,6 +353,53 @@ paths:
       tags:
         - Datasets
 
+  /sample/upload:
+    post:
+      summary: Upload a MoH sample JSON file for ingest
+      description: |
+        Upload a MoH schema (MoHSchemaV3) JSON file for sample registration ingestion.
+        A prefix must be provided to prepend to each donor's program_id when
```
Comment (daisieh) [inline schema.yml]: Since the prefix is a required parameter, it might be more clear to say something like "The prefix will be prepended to each donor's program_id"

File: schema.yml [hunk 2]
```diff
@@ -318,6 +353,53 @@ paths:
       tags:
         - Datasets
 
+  /sample/upload:
```
Comment (daisieh) [inline schema.yml]: This should be an endpoint under `datasets`, e.g. `/datasets/{dataset_id}/samples/upload`

**Review by daisieh — CHANGES_REQUESTED** [review]
I made a few comments inline on the schema, but I'm still not completely satisfied with the way we've set up the REST API on this. If you can ingest an omop file that contains multiple datasets, it doesn't make any sense for the upload to be a sub-path of `datasets`. It also doesn't make any sense to have a separate endpoint for the omop ingest and the sample ingest, since the daemon figures out which file type it is anyway. The two upload endpoints should be consolidated into a single one; is it possible for a POST to accept both files as part of the upload?

File: schema.yml (hunk 1 above)
Comment (SonQBChau) [reply schema.yml]: fixed

File: schema.yml (hunk 2 above)
Comment (SonQBChau) [reply schema.yml]: Since the moh json contains multiple datasets, this will allow only samples from 1 dataset to be uploaded at a time. Do we want that?

**Comment by SonQBChau** [comment]
> I made a few comments inline on the schema, but I'm still not completely satisfied with the way we've set up the REST API on this. If you can ingest an omop file that contains multiple datasets, it doesn't make any sense for the upload to be a sub-path of `datasets`. It also doesn't make any sense to have a separate endpoint for the omop ingest and the sample ingest, since the daemon figures out which file type it is anyway. The two upload endpoints should be consolidated into a single one; is it possible for a POST to accept both files as part of the upload?

A new upload sample endpoint is a quick way to ingest without redesigning the old one. Combine 2 would introduce many edge cases (required params, missing files, upload order, queued file response and error, large upload ..etc). Since time is short, I prefer not to modify the current flow and all the related tests

The better fix would be the ETL includes all the samples in the omop ingest, rather than doing two separate upload files

File: schema.yml (hunk 2 above)
Comment (daisieh) [reply schema.yml]: Isn't that true for the omop file as well?

**Comment by daisieh** [comment]
That's reasonable. I think I can approve this PR for now but will write a ticket for unifying the ingest flow later.

File: schema.yml (hunk 2 above)
Comment (SonQBChau) [reply schema.yml]: both upload endpoints accept multiple datasets

Comment (daisieh) [reply schema.yml]: Okay then it should be under the same endpoint as the other: 
```suggestion
  /datasets/upload/samples:
```

**Review by daisieh — APPROVED** [review]
One change to the endpoint syntax, then I think this is okay.

**Review by daisieh — CHANGES_REQUESTED** [review]
Actually wait a second: this is the result I got from the test ingest:
```
{
  "status": "Partial Success",
  "file_name": "input_mohccn_data.json",
  "file_size": 107669,
  "uploaded_at": "2026-03-05 20:02:29 UTC",
  "prefix": "SITE_PM2C",
  "data_type": "samples",
  "processed_at": "2026-03-05 20:02:29 UTC",
  "ingested_count": 21,
  "ingested_items": [
    "SITE_PM2C~SYNTH_01~SAMPLE_0001",
    "SITE_PM2C~SYNTH_01~SAMPLE_0002",
    "SITE_PM2C~SYNTH_01~SAMPLE_0003",
    "SITE_PM2C~SYNTH_01~SAMPLE_0004",
    "SITE_PM2C~SYNTH_01~SAMPLE_0005",
    "SITE_PM2C~SYNTH_01~SAMPLE_0006",
    "SITE_PM2C~SYNTH_01~SAMPLE_0007",
    "SITE_PM2C~SYNTH_01~SAMPLE_0008",
    "SITE_PM2C~SYNTH_01~SAMPLE_0009",
    "SITE_PM2C~SYNTH_01~SAMPLE_ALL_0001",
    "SITE_PM2C~SYNTH_01~SAMPLE_ALL_0002",
    "SITE_PM2C~SYNTH_01~SAMPLE_ALL_0003",
    "SITE_PM2C~SYNTH_01~SAMPLE_NULL_0003",
    "SITE_PM2C~SYNTH_01~SAMPLE_NULL_0001",
    "SITE_PM2C~SYNTH_01~SAMPLE_NULL_0002",
    "SITE_PM2C~SYNTH_01~SAMPLE_0010",
    "SITE_PM2C~SYNTH_01~SAMPLE_0011",
    "SITE_PM2C~SYNTH_01~SAMPLE_0012",
    "SITE_PM2C~SYNTH_01~SAMPLE_0013",
    "SITE_PM2C~SYNTH_01~SAMPLE_0014",
    "SITE_PM2C~SYNTH_01~SAMPLE_0015"
  ],
  "errors": [
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0017': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0006'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0018': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0006'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0016': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0006'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0019': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0007'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0020': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0007'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0021': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0007'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0022': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0008'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0023': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0008'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0024': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0008'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0027': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0009'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0025': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0009'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0026': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0009'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0028': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0010'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0029': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0010'.",
    "Error Sample 'SITE_PM2C~SYNTH_02~SAMPLE_0030': Person Not Found - No person found with person_source_value='SITE_PM2C~SYNTH_02~DONOR_0010'."
  ]
}
```
Why couldn't it find the SYNTH_02 donors?

**Comment by SonQBChau** [comment]
all the SYNTH_02 donors start from 21 in the test set. The input_mohccn_data.json Marion gave me is for small dataset

**Comment by daisieh** [comment]
Okay then, I think this is good to merge.
